src/common/utils/validators.py

```diff
@@ -35,125 +33,157 @@ def is_blueprint_instance_of(
     return False
 
 
-def _recursive_validate_single_attribute(
-    attribute: BlueprintAttribute, value: dict | bool | int | float | str, get_blueprint: Callable, key: str
-):
-    debug_message = f"Location: Entity in key '{key}'"
-    if attribute.is_primitive:
-        python_type = BuiltinDataTypes(attribute.attribute_type).to_py_type()
-        if not isinstance(value, python_type):
-            raise ValidationException(
-                f"Attribute '{attribute.name}' should be type '{python_type.__name__}'. Got '{type(value).__name__}'",
-                debug=debug_message,
-            )
-        if not attribute.is_optional:
-            if python_type == str and value == "":
-                raise ValidationException(f"Missing required attribute '{attribute.name}'", debug=debug_message)
+def _validate_primitive_attribute(attribute: BlueprintAttribute, value: bool | int | float | str, key: str):
+    python_type = BuiltinDataTypes(attribute.attribute_type).to_py_type()
+    if type(value) != python_type:
+        raise ValidationException(
+            f"Attribute '{attribute.name}' should be type '{python_type.__name__}'. Got '{type(value).__name__}'",
+            debug=_get_debug_message(key),
+        )
+    if not attribute.is_optional and python_type == str and value == "":
+        raise ValidationException(f"Missing required attribute '{attribute.name}'", debug=_get_debug_message(key))
 
-        return
-    if not attribute.is_optional and not value:
-        raise ValidationException(f"Missing required attribute '{attribute.name}'", debug=debug_message)
 
-    # Optional empty values are valid
-    if not value:
+def validate_entity_against_self(entity: dict | list, get_blueprint: Callable[..., Blueprint], key: str = "^") -> None:
```

Comment:
looks like validate_entity_against_self() and validate_entity() are very similar. Are there any guidelines when one should be used over the other?

Reply:
I though it better with two functions, than one function with many optional input variables. Basically, use validate-entity_against_self if you want to check that the entity is correct according to its own type. If you want to check the entity towards the parent's attribute type or a minimum pattern, use validate_entity. You could use the latter function to solve the former issue if you wanted, but that would make the function call far more verbose.

Reply:
alright, cool!